src/python/dart/service/pending_actions_check.py

```diff
@@ -0,0 +1,83 @@
+from dart.context.locator import injectable
+from dart.model.workflow import WorkflowState, WorkflowInstanceState
+from dart.model.action import ActionState
+
+import logging
+import boto3
+
+_logger = logging.getLogger(__name__)
+
+@injectable
+class PendingActionsCheck(object):
+    def __init__(self, action_service):
+        self._action_service = action_service
+        self._batch_client = boto3.client('batch')
+
+    def get_not_completed_workflow_instances(self, workflow_id, workflow_service):
+        wf = workflow_service.get_workflow(workflow_id, raise_when_missing=False)
+        if not wf:
+            _logger.info('Zombie Check: workflow (id={wf_id}) not found. log-info: {log_info}'.
+                         format(wf_id=workflow_id, log_info=workflow_msg.get('log_info')))
+            return None
+
+        if wf.data.state != WorkflowState.ACTIVE:
+            _logger.info('Zombie Check: expected workflow (id={wf_id}) to be in ACTIVE state. log-info: {log_info}'.
+                         format(wf_id=workflow_id, log_info=workflow_msg.get('log_info')))
+
+        # get all workflow_instances of current workflow:
+        NOT_COMPLETE_STATES = ['QUEUED', 'RUNNING']
+        all_wf_instances = workflow_service.find_workflow_instances(workflow_id)
+        current_wf_instances = [wf for wf in all_wf_instances if wf.data.state in NOT_COMPLETE_STATES]
+        _logger.info('Zombie Check: Found workflow instance ids (workflow_id={0}) instances = {1}'.format(workflow_id, current_wf_instances))
+
+        return current_wf_instances
+
+    def get_instance_actions(self, current_wf_instances):
+        # get all actions of not completed workflow_instances
+        incomplete_actions = []
+        action_2_wf_instance = {}
+        for wf_instance in current_wf_instances:
+            wf_instance_actions = self._action_service.find_actions(workflow_instance_id=wf_instance.id)
+            incomplete_actions.extend(wf_instance_actions)
+            for action in wf_instance_actions:
+                action_2_wf_instance[action.id] = wf_instance
+
+        jobs_2_actions = {}
+        for action in incomplete_actions:
+            if action.data.batch_job_id:
+                jobs_2_actions[action.data.batch_job_id] = action
+
+        return incomplete_actions, jobs_2_actions, action_2_wf_instance
+
+    def handle_done_batch_jobs_with_not_complete_wf_instances(self, batch_jobs, jobs_2_actions, action_2_wf_instance, workflow_service):
+        for job in batch_jobs.get('jobs'):
```

Comment:
this prob should be batch_jobs.get('jobs', []) to avoid trying to iterate over null

you use get quite a bit, which will by default return null if the key is not present.

Reply:
from http://boto3.readthedocs.io/en/latest/reference/services/batch.html#Batch.Client.describe_jobs there is always jobs key in the result and boto's standard is to have [] instead of null if none exits.
But we add this just in case.